### PR TITLE
Update `mkIterApp`, `mkIterInst` and `mkIterTyApp`

### PR DESCRIPTION
**Pull request by zliu41**
Changed `PlutusCore.MkPlc.mkIterApp`, `PlutusCore.MkPlc.mkIterInst` and `PlutusCore.MkPlc.mkIterTyApp`  to require an annotation to be provided for each argument. 

Also added `PlutusCore.MkPlc.mkIterAppNoAnn`, `PlutusCore.MkPlc.mkIterInstNoAnn` and `PlutusCore.MkPlc.mkIterTyAppNoAnn`.

### CODE REVIEW

**Comment by zliu41** [comment]
I think `fillAppContext` should also live in `MkPlc.hs` and be made polymorphic in `Term`, but I'm not doing that in this PR.

File: plutus-core/plutus-core/src/PlutusCore/MkPlc.hs [hunk 1]
```diff
@@ -210,23 +213,39 @@ mkImmediateTyAbs
...
     :: TermLike term tyname name uni fun
-    => ann
-    -> term ann -- ^ @f@
-    -> [term ann] -- ^@[ x0 ... xn ]@
-    -> term ann -- ^ @[f x0 ... xn ]@
-mkIterApp ann = foldl' (apply ann)
+    => term ann
+    -> [(ann, term ann)]
+    -> term ann
+mkIterApp = foldl' (uncurry . flip apply)
```
Comment (effectfully) [inline plutus-core/plutus-core/src/PlutusCore/MkPlc.hs:223]: Nit: at that point I'd add some names, `uncurry`, `.` and `flip` together is a bit too much even for a point-free abuser like myself.

File: plutus-core/plutus-core/examples/PlutusCore/Examples/Data/Vec.hs [hunk 2]
```diff
@@ -334,7 +334,7 @@ scottHead = runQuote $ do
         . TyAbs () a (Type ())
         . TyAbs () n natK
         . LamAbs () xs (mkIterTyApp () scottVec [TyVar () a, TyApp () succT $ TyVar () n])
-        $ mkIterApp ()
+        $ mkIterAppNoAnn
```
Comment (effectfully) [inline plutus-core/plutus-core/examples/PlutusCore/Examples/Data/Vec.hs:337]: `mkIterTyApp` shares the same annotation, `mkIterApp` doesn't and we have a separate `mkIterAppNoAnn`. Looks kinda messy. It also messes with one's reflexes: when creating an AST, one uses an explicit annotation for virtually every function/constructor except `mkIterApp` and `mkIterInst` all of a sudden.

**Review by effectfully — APPROVED** [review]
This does look like an improvement for me, but I'm feeling something a bit unsettling deep inside my soul about the new API.

File: plutus-core/plutus-core/src/PlutusCore/MkPlc.hs (hunk 1 above)
Comment (zliu41) [reply plutus-core/plutus-core/src/PlutusCore/MkPlc.hs:223]: Understandable. I find it easier to read though :smile:

File: plutus-core/plutus-core/examples/PlutusCore/Examples/Data/Vec.hs (hunk 2 above)
Comment (zliu41) [reply plutus-core/plutus-core/examples/PlutusCore/Examples/Data/Vec.hs:337]: Indeed, the same change should be made to `mkIterTyApp`, which I overlooked.

**Review by zliu41 — COMMENTED** [review]
> This does look like an improvement for me, but I'm feeling something a bit unsettling deep inside my soul about the new API.

I think sharing a non-unit annotation when constructing nested Apply or TyInst nodes is, in most cases, wrong (e.g., what the parsers are doing currently).

File: plutus-core/plutus-core/src/PlutusCore/MkPlc.hs (hunk 1 above)
Comment (michaelpj) [reply plutus-core/plutus-core/src/PlutusCore/MkPlc.hs]: > I find it easier to read though

I'm kind of amazed if this is true :D

File: plutus-core/plutus-core/src/PlutusCore/Parser.hs [hunk 3]
```diff
@@ -46,7 +47,8 @@ lamTerm = withSpan $ \sp ->
 
 appTerm :: Parser PTerm
 appTerm = withSpan $ \sp ->
-    inBrackets $ mkIterApp sp <$> term <*> some term
+    -- TODO: should not use the same `sp` for all arguments.
```
Comment (michaelpj) [inline plutus-core/plutus-core/src/PlutusCore/Parser.hs:50]: Not clear to me what it _should_ be. Do you want something like this?
```
[ f a b c ]
--> 
[[ f a ](0:5) b](0:7) c ](0:11)
```

File: plutus-core/plutus-ir/src/PlutusIR/Compiler/Datatype.hs [hunk 4]
```diff
@@ -590,7 +591,10 @@ compileDatatype r body d = do
         vars = fmap PIR.defVar constrDefs ++ [ PIR.defVar destrDef ]
         vals = fmap PIR.defVal constrDefs ++ [ PIR.defVal destrDef ]
     -- See note [Abstract data types]
-    pure $ PIR.mkIterApp p (PIR.mkIterInst p (PIR.mkIterTyAbs tyVars (PIR.mkIterLamAbs vars body)) tys) vals
+    pure $
+      PIR.mkIterApp
+        (PIR.mkIterInst (PIR.mkIterTyAbs tyVars (PIR.mkIterLamAbs vars body)) ((p,) <$> tys))
```
Comment (michaelpj) [inline plutus-core/plutus-ir/src/PlutusIR/Compiler/Datatype.hs:596]: I mean, this stuff reads much worse now but :shrug:

**Review by michaelpj — COMMENTED** [review]
I'm not entirely convinced this is an improvement. We could have fixed the parsers without changing all the rest, and it seems to me that in all the other cases which were changed we are _not_ going to do anything different than using a single annotation for all of it. So we've made 90% of the usages worse for no benefit, when we could just have special-cased the ones that did benefit.

File: plutus-core/plutus-core/src/PlutusCore/MkPlc.hs (hunk 1 above)
Comment (zliu41) [reply plutus-core/plutus-core/src/PlutusCore/MkPlc.hs]: I mean, `uncurry` and `flip` are just restructuring the arguments and can be ignored, so it's really just `fold' apply`.

File: plutus-core/plutus-core/src/PlutusCore/Parser.hs (hunk 3 above)
Comment (zliu41) [reply plutus-core/plutus-core/src/PlutusCore/Parser.hs:50]: Yes, that's what I had in mind.

**Review by zliu41 — COMMENTED** [review]
> We could have fixed the parsers without changing all the rest

I need the new `mkIterApp` in #5323. We can keep both versions, but I don't know how to name them. Am not a huge fan of either `mkIterApp'` or `mkIterAppSharedAnn`.

**Comment by kwxm** [comment]
@zliu41 Should we be doing the same thing for the UPLC of `mkIterApp` [here](https://github.com/input-output-hk/plutus/blob/e2f664e52a5cfef99bf863de945e9f03d505b071/plutus-core/untyped-plutus-core/src/UntypedPlutusCore/MkUPlc.hs#L16)?

**Comment by zliu41** [comment]
> Should we be doing the same thing for the [UPLC version of mkIterApp](https://github.com/input-output-hk/plutus/blob/e2f664e52a5cfef99bf863de945e9f03d505b071/plutus-core/untyped-plutus-core/src/UntypedPlutusCore/MkUPlc.hs#L16)?

Yes, it should be deleted. The other `mkIterApp` is polymorphic so we don't need the UPLC version.

**Comment by zliu41** [comment]
It is in fact deleted in https://github.com/input-output-hk/plutus/pull/5323.

**Pull request merged**